Signal/src/ViewControllers/ConversationView/Cells/ConversationViewCell.h

```diff
@@ -0,0 +1,63 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+NS_ASSUME_NONNULL_BEGIN
+
+@class ConversationViewCell;
+@class ConversationViewItem;
+@class OWSContactOffersInteraction;
+@class TSAttachmentPointer;
+@class TSAttachmentStream;
+@class TSInteraction;
+@class TSMessage;
+@class TSOutgoingMessage;
+
+@protocol ConversationViewCellDelegate <NSObject>
+
+- (void)didTapImageViewItem:(ConversationViewItem *)viewItem
+           attachmentStream:(TSAttachmentStream *)attachmentStream
+                  imageView:(UIView *)imageView;
+- (void)didTapVideoViewItem:(ConversationViewItem *)viewItem attachmentStream:(TSAttachmentStream *)attachmentStream;
+- (void)didTapAudioViewItem:(ConversationViewItem *)viewItem attachmentStream:(TSAttachmentStream *)attachmentStream;
+- (void)didTapOversizeTextMessage:(NSString *)displayableText attachmentStream:(TSAttachmentStream *)attachmentStream;
+- (void)didTapFailedIncomingAttachment:(ConversationViewItem *)viewItem
+                     attachmentPointer:(TSAttachmentPointer *)attachmentPointer;
+- (void)didTapFailedOutgoingMessage:(TSOutgoingMessage *)message;
+
+- (void)showMetadataViewForMessage:(TSMessage *)message;
+
+#pragma mark - System Cell
+
+// TODO: We might want to decompose this method.
+- (void)didTapSystemMessageWithInteraction:(TSInteraction *)interaction;
+- (void)didLongPressSystemMessageCell:(ConversationViewCell *)systemMessageCell fromView:(UIView *)fromView;
+
+#pragma mark - Offers
+
+- (void)tappedUnknownContactBlockOfferMessage:(OWSContactOffersInteraction *)interaction;
+- (void)tappedAddToContactsOfferMessage:(OWSContactOffersInteraction *)interaction;
+- (void)tappedAddToProfileWhitelistOfferMessage:(OWSContactOffersInteraction *)interaction;
+
+@end
+
+#pragma mark -
+
+@interface ConversationViewCell : UICollectionViewCell
+
+@property (nonatomic, nullable, weak) id<ConversationViewCellDelegate> delegate;
+
+@property (nonatomic, nullable) ConversationViewItem *viewItem;
+
+@property (nonatomic) BOOL isCellVisible;
```

Comment:
It's TODO but we need to go back to being aggressive about unload view state when cells are offscreen.  We used to do this and its key for performance.

Reply:
It might be simpler to turn off `UICollectionView`'s cell prefetching.  I haven't decided yet.

Reply:
This work will include carefully caching a subset of cell state on the view items - but we need to be very careful to clear the cached state for cells that are sufficiently far offscreen.

Reply:
> It might be simpler to turn off UICollectionView's cell prefetching. I haven't decided yet.

But don't we *want* to take advantage of prefetching? It's one of the things that can help make scrolling smoother.

Reply:
Right.  It's a trade-off between:

* Simpler code but worse perf for textual messages with no pre-fetching.
* More complicated code but better perf with pre-fetching + unload offscreen media + caching offscreen media within a certain window.